text/differential-layer-format.md

```diff
@@ -0,0 +1,143 @@
+
+- FPI, 0 revisions
+- will-init WAL record, 0 revisions
+- FPI-initiated branch, N WAL records depending on that FPI
+- WAL-record initiated branch, N WAL records depending on that WAL record
+- ...
+
+In the physical format, because the LSNs of the WAL records are sorted and stored
+in increasing order, we could apply compression to this array to reduce the
+storage overhead of the full branch structure.
+
+Additionally, because the length of the WAL record is stored in it's header, we
+don't actually need to store the length of the WAL record in the branch structure;
+we can read it directly from the header, saving 4 bytes per WAL record.
+
+With LSN array compression and the emission of the WAL record length also being
+stored outside the WAL record, we can reduce the storage overhead of non-FPI WAL
+records from 28 bytes per record down to less than 8 bytes per record, with an
+expected value of ~ 4 bytes (depending on how often a page is modified; more
+frequent / less noizy WAL record generation = lower overhead).
+
+
+
+### Page versions array
+The Delta layer uses an in-memory sorted array of <lsn, pageversion> pairs to
+represent the changes to a page. This is correct and allows for O(log N) version
+lookups (for N = number of total page versions). This is OK, but can be improved
+upon: as mentioned in "Storage and file formats; Structures", we always
+need a base page image to reconstruct a page. By limiting the array of page
+versions stored for each page in the Layer to Branches, we eliminate a
+significant amount of entries from the search space; improving this search speed
+by O(log fraction of non-image page-versions) and reducing memory usage.
+
+Because we generally only require the latest version of a page, we store the newest
+versions of the page in the front of the array, so that the access to this data
+requires few extra operations.
+
+Assuming that layer files will not grow past a certain size (2^48 - 1 bytes),
+we can can use an 48-bit integer to store the location of a Branch.
+When in memory, with the 8-byte LSN, we have 2 bytes left to store some more
+metadata of the branch in the BranchPtr, such as type info and other relevant
+information.
+
+
+The per-page data layout of the Differential layer will thus be:
+
+```text
+  Page data:
+    ... metadata
+    page_versions: sorted array of (Lsn -> BranchPtr) pairs, in descending order of LSN
+      Lsn -> BranchPtr:
+        Lsn: 8 bytes
+        BranchPtr: 8 bytes
+          offset_to_branch: 6 bytes
+          branch type specific info: 2 bytes
+          ...
+      Lsn -> BranchPtr -> ...
+      Lsn -> BranchPtr -> ...
+      Lsn -> BranchPtr -> ...
+      ...
+```
+
+Note that the definition of BranchPtr is partially dependent on what data we
+want to store in the BranchPtr, and what data we want to store locally in the
+Branch. Either way, the overhead of one branch in the page versions array is
+16 bytes.
+
```

Comment:
What is the intended accessed pattern?
The obvious assumption is that compute node is not requesting recently updated pages.
So page server has some time to reconstruct them in background. It is now done by checkpointer,
but it may be better performed by separate thread to distinct persisting data and page reconstruction.
In the ideal case, page is reconstructed be checkpointer (reconstructor) before it is requested by get_page_at_lsn.
In this case latency of get_page_at_lsn will be minimal. And we do not need random access at all!
Checkpointer can sequentially read all data from delta layer and reconstruct page images. So there is no need to use some more sophisticated data format to provide to the branch. 

The main question is what to do with reconstructed images? Storing them in image layer may cause write amplification. If them are stored in page cache, then eviction algorithm will make part of this reconstuction work useless, because reconstructed page is evicted before it is accessed. I do not have good answer for this question.
May be we can really do not perform eager page materialization and always do it only on demand. But in this case reuestion page from pageserver will never be fast enough.